src/accounts_refactor.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Seahorse library.
+
+// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The accounts module.
+//!
+//! This module defines [Account], [AccountEditor], and [Accounts], which provide CURD (create, read,
+//! update, and delete) operations, with the use of [KeyValueStore] to control the accounts resource.
+
+use crate::{
+    events::{EventSource, LedgerEvent},
+    key_scan::BackgroundKeyScan,
+    key_value_store::KeyValueStore,
+    txn_builder::RecordInfo,
+    KeystoreError,
+};
+use chrono::{DateTime, Local};
+use jf_cap::{
+    keys::{FreezerKeyPair, FreezerPubKey, UserAddress, UserKeyPair, ViewerKeyPair, ViewerPubKey},
+    structs::AssetCode,
+    MerkleCommitment,
+};
+use primitive_types::U256;
+use reef::Ledger;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::{Deref, DerefMut};
+
+/// Keys with a public part.
+pub trait KeyPair: Clone + Send + Sync {
+    type PubKey: Clone + DeserializeOwned + Hash + Eq + Serialize;
+    fn pub_key(&self) -> Self::PubKey;
+}
+
+impl KeyPair for ViewerKeyPair {
+    type PubKey = ViewerPubKey;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.pub_key()
+    }
+}
+
+impl KeyPair for FreezerKeyPair {
+    type PubKey = FreezerPubKey;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.pub_key()
+    }
+}
+
+impl KeyPair for UserKeyPair {
+    // The PubKey here is supposed to be a conceptual "primary key" for looking up UserKeyPairs. We
+    // typically want to look up UserKeyPairs by Address, not PubKey, because if we have a PubKey we
+    // can always get an Address to do the lookup.
+    type PubKey = UserAddress;
+
+    fn pub_key(&self) -> Self::PubKey {
+        self.address()
+    }
+}
+
+#[derive(Clone, Deserialize, Serialize)]
+#[serde(bound = "Key: DeserializeOwned + Serialize")]
+pub struct Account<L: Ledger, Key: KeyPair> {
+    /// The account key.
+    key: Key,
+    /// The account description.
+    pub description: String,
+    /// Whether the account is used.
+    pub used: bool,
+    /// Optional ledger scan.
+    scan: Option<BackgroundKeyScan<L>>,
+    /// The list of asset codes of the account.
+    assets: Vec<AssetCode>,
```

Comment:
Similar question as with `records`. This is going to change dynamically depending on which records  we own at any given time, so duplicating it here risks things getting out of sync. We might want to consider replacing this field with a controller-level function that iterates over unspent records, filters for those that are owned/viewable/freezable by this account, and collects the corresponding assets